### PR TITLE
Add dependency resolution to ContentDB

### LINKED ISSUE #8891: Add dependency installation to ContentDB
Automatically install dependencies for a mod when installing it.

Alternative to #8661

## How it works

1. User decides to install Technic
2. A request is made to a ContentDB API to determine the dependencies.
   The API returns a list of dependencies and the packages which fulfil it.

   Example: https://content.minetest.net/api/packages/RealBadAngel/technic/dependencies/

3. Minetest works out what should be installed, based on the currently installed mods. Minetest will check in both the global mod location and the mod location of the currently selected game. It will prefer to use local content, and it will prefer to not install full games from ContentDB.
4. A simple dependency information dialog is shown, to tell the user what will be installed.
5. (optional) The user may click "Edit" to change which mods are to be installed
6. For each mod to be installed, start again from step 1.
7. Add all mods to the download manager
8. Download

## Milestones

- [ ] Add download manager to queue and dispatch installs.
- [ ] Read `/api/packages/<author>/<name>/dependencies/` from ContentDB
- [ ] Add simple dependency information dialog, to show what is to be installed.
- [ ] Add resolve dependency dialog, to allow the user to choose optional mods and alternatives

## Mock Ups

Simple mode

image

Advanced mode

image

### PR DESCRIPTION
**Pull request by rubenwardy**
Fixes #8891

image

image

image

## How does it work?

When installing a package, an API request is made to ContentDB to find out the dependencies. If there are no dependencies, then the mod is installed straight away.

If there are dependencies, it will resolve them and show a dialog with a list of mods to install.

Resolving works like this:

1. Fetch dependencies for the package from ContentDB's API
2. For each hard dependency:
	1. Check current game, exit if dep found
	2. Check installed mods, exit if found
    3. Check available mods from ContentDB:
		1. Choose a package to install. Prefer higher scores and exact base name matches (rather than modpacks).
		2. Resolve dependencies for this package - ie, goto 1.

The ContentDB API is a dictionary of packages to dependencies. The dictionary will allow ContentDB to prefetch dependencies without any client changes. For example, say you request the dependencies for Mobs Monster. It's pretty likely that the next request you'll make is for Mobs Redo, and so ContentDB can prevent the need for another request by including the dependency information for Mobs Redo in the response for Mobs Monster.

Example API response:

```json
{
  "TenPlus1/mobs_monster": [
    {
      "is_optional": false,
      "name": "default",
      "packages": [
        "Casimir/voxelgarden",
        "Wuzzy/tutorial",
        "bell07/whynot_game",
        "Minetest/minetest_game"
      ]
    },
    {
      "is_optional": false,
      "name": "mobs",
      "packages": [
        "TenPlus1/mobs",
        "bell07/whynot_game",
        "niwla23/mt_gme_improved"
      ]
    }
  ],
  "TenPlus1/mobs": [
    {
      "is_optional": false,
      "name": "default",
      "packages": [
        "Casimir/voxelgarden",
        "Wuzzy/tutorial",
        "DELTA_FORCE/terrariaswords",
        "Minetest/minetest_game"
      ]
    }
  ]
}
```

## To do

This PR requires #9995 

- [x] Check installed mods
- [x] Add ability to not install any dependencies
- [x] Testing

## How to test

Todo

### CODE REVIEW

File: builtin/mainmenu/dlg_contentstore.lua [hunk 1]
```diff
@@ -123,6 +139,290 @@ local function start_install(calling_dialog, package)
...
+	local tmpdir = os.tempfolder()
+	local target = tmpdir .. DIR_DELIM .. "package_deps.json"
+
+	assert(core.create_dir(tmpdir))
+
+	local url_fmt = "/api/packages/%s/dependencies/?protocol_version=%s&engine_version=%s&only_hard=1"
+	local version = core.get_version()
+	local base_url = core.settings:get("contentdb_url")
+	local url = base_url .. url_fmt:format(package.id, core.get_max_supp_proto(), version.string)
+	core.download_file(url, target)
```
Comment (sfan5) [inline builtin/mainmenu/dlg_contentstore.lua:168]: does the engine not have a method to download an url into a string?

Comment (rubenwardy) [reply builtin/mainmenu/dlg_contentstore.lua:168]: It does not in the menu API, but does in the server-side API. Perhaps I shou I could add it to the menu API?

File: builtin/mainmenu/dlg_contentstore.lua [hunk 2]
```diff
@@ -123,6 +139,290 @@ local function start_install(calling_dialog, package)
...
+		package.queued = true
+	end
+end
+
+local function get_raw_dependencies(package)
+	if package.raw_deps then
+		return package.raw_deps
+	end
+
+	print("#### Fetching raw deps for " .. package.id)
```
Comment (rubenwardy) [inline builtin/mainmenu/dlg_contentstore.lua:157]: remove debug print

**Comment by rubenwardy** [comment]
This is no longer WIP, but it does require #9995 to be merged first

**Comment by rubenwardy** [comment]
Updated screenshot:

image

**Comment by rubenwardy** [comment]
closing until other PR is merged

**Comment by rubenwardy** [comment]
Note: there's currently a bug in ContentDB (the website) where it incorrectly lists mods as dependencies even if the package contains that mods. This will cause games to show errors

**Comment by LoneWolfHT** [comment]
Added this to my client. If there are any issues I'll let you know

**Comment by rubenwardy** [comment]
> Added this to my client. If there are any issues I'll let you know

Did you find any issues? Have you used it much?

**Comment by LoneWolfHT** [comment]
> > Added this to my client. If there are any issues I'll let you know
> 
> Did you find any issues? Have you used it much?

A few times. So far no issues at all. Works quite well

**Review by sfan5 — APPROVED** [review]

**Pull request merged**